core/src/main/scala/org/apache/spark/rdd/MapPartitionsWithPreparationRDD.scala

```diff
@@ -38,12 +38,25 @@ private[spark] class MapPartitionsWithPreparationRDD[U: ClassTag, T: ClassTag, M
 
   override def getPartitions: Array[Partition] = firstParent[T].partitions
 
+  private[this] var preparedArgument: Option[M] = None
+
+  def prepare(): Unit = {
+    // This could be called multiple times
+    if (preparedArgument.isEmpty) {
+      preparedArgument = Some(preparePartition())
+    }
+  }
+
   /**
    * Prepare a partition before computing it from its parent.
    */
   override def compute(partition: Partition, context: TaskContext): Iterator[U] = {
-    val preparedArgument = preparePartition()
+    prepare()
+    // The same RDD could be called multiple times in one task, each call of compute() should
+    // have separate prepared argument.
+    val prepared = preparedArgument.get
+    preparedArgument = None
```

Comment:
By the way, you can rewrite this without the `get` if you make `prepare()` return the argument:
```
val preparedArg = prepare()
preparedArgument = None
```